baybe/objectives/desirability.py

```diff
@@ -0,0 +1,135 @@
+
+    Raises:
+        ValueError: If any of the weights is non-positive.
+
+    Returns:
+        The normalized weights.
+    """
+    array = np.asarray(cattrs.structure(weights, tuple[float, ...]))
+    if not np.all(array > 0.0):
+        raise ValueError("All weights must be strictly positive.")
+    return tuple(array / array.sum())
+
+
+def _is_all_numerical_targets(
+    x: tuple[Target, ...], /
+) -> TypeGuard[tuple[NumericalTarget, ...]]:
+    """Typeguard helper function."""
+    return all(isinstance(y, NumericalTarget) for y in x)
+
+
+def scalarize(
+    values: npt.ArrayLike, scalarization: Scalarization, weights: npt.ArrayLike
+) -> np.ndarray:
+    """Scalarize the rows of a 2-D array, producing a 1-D array.
+
+    Args:
+        values: The 2-D array whose rows are to be scalarized.
+        scalarization: The scalarization mechanism to be used.
+        weights: Weights for the columns of the input array.
+
+    Raises:
+        ValueError: If the requested scalarization mechanism is not implemented.
+
+    Returns:
+        np.ndarray: A 1-D array containing the scalarized values.
+    """
+    func: Callable
+    if scalarization is Scalarization.GEOM_MEAN:
+        func = geom_mean
+    elif scalarization is Scalarization.MEAN:
+        func = partial(np.average, axis=1)
+    else:
+        raise NotImplementedError(
+            f"No scalarization mechanism defined for '{scalarization.name}'."
+        )
+    return func(values, weights=weights)
+
+
+@define(frozen=True)
+class DesirabilityObjective(Objective):
+    """An objective scalarizing multiple targets using desirability values."""
+
+    targets: tuple[Target, ...] = field(
+        converter=to_tuple,
+        validator=[min_len(2), deep_iterable(member_validator=instance_of(Target))],  # type: ignore[type-abstract]
+    )
+    "The targets considered by the objective."
+
+    weights: tuple[float, ...] = field(converter=_normalize_weights)
+    """The weights to balance the different targets.
+    By default, all targets are considered equally important."""
+
+    scalarization: Scalarization = field(
+        default=Scalarization.GEOM_MEAN, converter=Scalarization
+    )
+    """The mechanism to scalarize the weighted desirability values of all targets."""
+
+    @weights.default
+    def _default_weights(self) -> tuple[float, ...]:
+        """Create unit weights for all targets."""
+        return tuple(1.0 for _ in range(len(self.targets)))
+
+    @targets.validator
+    def _validate_targets(self, _, targets) -> None:  # noqa: DOC101, DOC103
+        if not _is_all_numerical_targets(targets):
+            raise ValueError(
+                f"'{self.__class__.__name__}' currently only supports targets "
+                f"of type '{NumericalTarget.__name__}'."
+            )
+        if not all(target.is_normalized for target in targets):
+            raise ValueError(
+                "All targets must be normalized, which requires setting appropriate "
```

Comment:
the wording iin this error and `is_normalized` are a bit weird to me. `is normalized` I would associate with the respective values I fed in, but you mean it as a proeprty of the target object, maybe hteres a better way of saying that

Reply:
Yeah, I was struggling with this a lot and would appreciate any suggestions here. But you do get my reason for the change, right? Even though I'm one of the devs, it was not obvious for me to recap what is the condition we are checking against and I felt that there was just too much implementation details in the wrong class that were not properly abstracted away. Felt like a place where things could break easily, hence the explicit property. But I'm more than happy to adjust namings, error messages, etc 👍🏼

Reply:
Hi @Scienfitz, let's try to get this bigger PR merged today. This one here is the only open blocker. Do you have any suggestions how to resolve best?

Reply:
sry i had written some suggestions before but didnt send the comen

what about `has_normalization`, `has_transformation`, `is_transformed` or so? This is minor so if you want to move on without change thats fine